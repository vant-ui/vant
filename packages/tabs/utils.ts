import { raf } from '../utils/dom/raf';

export function scrollLeftTo(el: HTMLElement, to: number, duration: number) {
  let count = 0;
  const from = el.scrollLeft;
  const frames = Math.round((duration * 1000) / 16);

  function animate() {
    el.scrollLeft += (to - from) / frames;

    if (++count < frames) {
      raf(animate);
    }
  }

  animate();
}
