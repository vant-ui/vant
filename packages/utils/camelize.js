const camelizeRE = /-(\w)/g;
export default function(str) {
  return str.replace(camelizeRE, (_, c) => c ? c.toUpperCase() : '');
}
