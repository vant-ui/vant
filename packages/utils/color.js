export const RED = '#f44';
export const BLUE = '#1989fa';
export const GREEN = '#4b0';
export const GRAY_DARK = '#969799';
