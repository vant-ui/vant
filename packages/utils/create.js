/**
 * Create a component with common options
 */
import '../locale';
import Icon from '../icon';
import i18n from '../mixins/i18n';
import install from './install';
import Loading from '../loading';

export default function(sfc) {
  sfc.install = sfc.install || install;
  sfc.mixins = sfc.mixins || [];
  sfc.mixins.push(i18n);
  sfc.components = sfc.components || {};
  sfc.components.icon = Icon;
  sfc.components.loading = Loading;

  return sfc;
};
