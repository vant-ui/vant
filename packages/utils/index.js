import Vue from 'vue';

export { use, useSlots } from './use';

export const isServer = Vue.prototype.$isServer;

export function noop() {}

export function isDef(value) {
  return value !== undefined && value !== null;
}

export function isObj(x) {
  const type = typeof x;
  return x !== null && (type === 'object' || type === 'function');
}

export function get(object, path) {
  const keys = path.split('.');
  let result = object;

  keys.forEach(key => {
    result = isDef(result[key]) ? result[key] : '';
  });

  return result;
}

const camelizeRE = /-(\w)/g;
export function camelize(str) {
  return str.replace(camelizeRE, (_, c) => c.toUpperCase());
}

export function isAndroid() {
  /* istanbul ignore next */
  return isServer ? false : /android/.test(navigator.userAgent.toLowerCase());
}

export function range(num, min, max) {
  return Math.min(Math.max(num, min), max);
}
