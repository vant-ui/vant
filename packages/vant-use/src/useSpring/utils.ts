import { SpringConfig } from './config';

export const now =
  typeof performance !== 'undefined'
    ? performance.now.bind(performance)
    : Date.now.bind(Date);

export const isArray = Array.isArray.bind(Array);

const reusedTuple: [number, number] = [0, 0];
export function stepper(
  secondPerFrame: number,
  x: number,
  v: number,
  destX: number,
  spring: SpringConfig,
): [number, number] {
  // Spring stiffness, in kg / s^2

  // for animations, destX is really spring length (spring at rest). initial
  // position is considered as the stretched/compressed position of a spring
  const Fspring = -spring.tension * (x - destX);

  // Damping, in kg / s
  const Fdamper = -spring.friction * v;

  const a = (Fspring + Fdamper) / (spring.mass || 1);

  const newV = v + a * secondPerFrame;
  const newX = x + newV * secondPerFrame;

  const precision = spring.precision || 0.01;

  if (Math.abs(newV) < precision && Math.abs(newX - destX) < precision) {
    reusedTuple[0] = destX;
    reusedTuple[1] = 0;
    return reusedTuple;
  }

  reusedTuple[0] = newX;
  reusedTuple[1] = newV;

  return reusedTuple;
}
