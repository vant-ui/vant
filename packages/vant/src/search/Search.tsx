import { ref, defineComponent, ExtractPropTypes } from 'vue';

// Utils
import {
  pick,
  extend,
  truthProp,
  preventDefault,
  makeStringProp,
  createNamespace,
} from '../utils';
import { fieldSharedProps } from '../field/Field';

// Composables
import { useExpose } from '../composables/use-expose';

// Components
import { Field, FieldInstance } from '../field';

// Types
import type { SearchShape } from './types';

const [name, bem, t] = createNamespace('search');

const props = extend({}, fieldSharedProps, {
  label: String,
  shape: makeStringProp<SearchShape>('square'),
  leftIcon: makeStringProp('search'),
  clearable: truthProp,
  actionText: String,
  background: String,
  showAction: Boolean,
});

export type SearchProps = ExtractPropTypes<typeof props>;

export default defineComponent({
  name,

  props,

  emits: ['search', 'cancel', 'update:modelValue'],

  setup(props, { emit, slots, attrs }) {
    const filedRef = ref<FieldInstance>();

    const onCancel = () => {
      if (!slots.action) {
        emit('update:modelValue', '');
        emit('cancel');
      }
    };

    const onKeypress = (event: KeyboardEvent) => {
      const ENTER_CODE = 13;
      if (event.keyCode === ENTER_CODE) {
        preventDefault(event);
        emit('search', props.modelValue);
      }
    };

    const renderLabel = () => {
      if (slots.label || props.label) {
        return (
          <label class={bem('label')} for={props.id}>
            {slots.label ? slots.label() : props.label}
          </label>
        );
      }
    };

    const renderAction = () => {
      if (props.showAction) {
        const text = props.actionText || t('cancel');
        return (
          <div
            class={bem('action')}
            role="button"
            tabindex={0}
            onClick={onCancel}
          >
            {slots.action ? slots.action() : text}
          </div>
        );
      }
    };

    const blur = () => filedRef.value?.blur();
    const focus = () => filedRef.value?.focus();

    const fieldPropNames = Object.keys(fieldSharedProps) as Array<
      keyof typeof fieldSharedProps
    >;

    const renderField = () => {
      const fieldAttrs = extend({}, attrs, pick(props, fieldPropNames));

      const onInput = (value: string) => emit('update:modelValue', value);

      return (
        <Field
          v-slots={pick(slots, ['left-icon', 'right-icon'])}
          ref={filedRef}
          type="search"
          class={bem('field')}
          border={false}
          onKeypress={onKeypress}
          onUpdate:modelValue={onInput}
          {...fieldAttrs}
        />
      );
    };

    useExpose({ focus, blur });

    return () => (
      <div
        class={bem({ 'show-action': props.showAction })}
        style={{ background: props.background }}
      >
        {slots.left?.()}
        <div class={bem('content', props.shape)}>
          {renderLabel()}
          {renderField()}
        </div>
        {renderAction()}
      </div>
    );
  },
});
