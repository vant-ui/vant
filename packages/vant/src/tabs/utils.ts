import { raf, cancelRaf } from '@vant/use';
import { ScrollElement, getScrollTop, setScrollTop } from '../utils';

let scrollLeftToRafId: number;
let scrollTopToRafId: number;

export function scrollLeftTo(
  scroller: HTMLElement,
  to: number,
  duration: number
) {
  cancelRaf(scrollLeftToRafId);

  let count = 0;
  const from = scroller.scrollLeft;
  const frames = duration === 0 ? 1 : Math.round((duration * 1000) / 16);

  function animate() {
    scroller.scrollLeft += (to - from) / frames;

    if (++count < frames) {
      scrollLeftToRafId = raf(animate);
    }
  }

  animate();
}

export function scrollTopTo(
  scroller: ScrollElement,
  to: number,
  duration: number,
  callback: () => void
) {
  cancelRaf(scrollTopToRafId);

  let current = getScrollTop(scroller);
  const isDown = current < to;
  const frames = duration === 0 ? 1 : Math.round((duration * 1000) / 16);
  const step = (to - current) / frames;

  function animate() {
    current += step;

    if ((isDown && current > to) || (!isDown && current < to)) {
      current = to;
    }

    setScrollTop(scroller, current);

    if ((isDown && current < to) || (!isDown && current > to)) {
      scrollTopToRafId = raf(animate);
    } else if (callback) {
      scrollTopToRafId = raf(callback as FrameRequestCallback);
    }
  }

  animate();
}
