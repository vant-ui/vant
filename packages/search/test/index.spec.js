import Search from '..';
import { mount } from '../../../test/utils';

test('listen input event', () => {
  const onInput = jest.fn();
  const wrapper = mount(Search, {
    context: {
      on: {
        input: onInput
      }
    }
  });

  const input = wrapper.find('input');
  input.element.value = '1';
  input.trigger('input');

  expect(onInput).toBeCalledWith('1');
});

test('cancel search', () => {
  const onInput = jest.fn();
  const onCancel = jest.fn();

  const wrapper = mount(Search, {
    propsData: {
      value: 'test',
      showAction: true
    },
    context: {
      on: {
        input: onInput,
        cancel: onCancel
      }
    }
  });

  const cancel = wrapper.find('.van-search__action div');
  cancel.trigger('click');

  expect(onInput).toBeCalledWith('');
  expect(onCancel).toBeCalled();
});

test('emit a search event', () => {
  const onSearch = jest.fn();
  const onKeypress = jest.fn();

  const wrapper = mount(Search, {
    context: {
      on: {
        search: onSearch,
        keypress: onKeypress
      }
    }
  });

  const input = wrapper.find('input');
  input.trigger('keypress.enter');
  input.trigger('keypress.a');

  expect(onSearch).toBeCalled();
  expect(onKeypress).toBeCalled();
});

test('render label slot', () => {
  const wrapper = mount(Search, {
    scopedSlots: {
      label() {
        return 'Custom Label';
      }
    }
  });

  expect(wrapper).toMatchSnapshot();
});
