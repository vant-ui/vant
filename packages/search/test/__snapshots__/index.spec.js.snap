// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders search correctly 1`] = `
<section class="van-doc-demo-section demo-search" style="background:undefined;">
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">基础用法</h2>
    <div class="van-search" style="background:#f2f2f2;">
      <i class="van-icon van-icon-search" style="color:undefined;">
        <!---->
      </i>
      <div placeholder="请输入商品名称" class="van-cell van-field van-field--has-icon">
        <!---->
        <!---->
        <div class="van-cell__value van-cell__value--alone">
          <input type="search" placeholder="请输入商品名称" value="" class="van-field__control">
          <!---->
          <div class="van-field__icon" style="display:none;">
            <i class="van-icon van-icon-clear" style="color:undefined;">
              <!---->
            </i>
          </div>
          <!---->
        </div>
        <!---->
      </div>
      <!---->
    </div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">监听对应事件</h2>
    <form action="/">
      <div class="van-search van-search--show-action" style="background:#f2f2f2;">
        <i class="van-icon van-icon-search" style="color:undefined;">
          <!---->
        </i>
        <div placeholder="请输入商品名称" class="van-cell van-field van-field--has-icon">
          <!---->
          <!---->
          <div class="van-cell__value van-cell__value--alone">
            <input type="search" placeholder="请输入商品名称" value="" class="van-field__control">
            <!---->
            <div class="van-field__icon" style="display:none;">
              <i class="van-icon van-icon-clear" style="color:undefined;">
                <!---->
              </i>
            </div>
            <!---->
          </div>
          <!---->
        </div>
        <div class="van-search__action">
          <div class="van-search__cancel">取消</div>
        </div>
      </div>
    </form>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">自定义行动按钮</h2>
    <div class="van-search van-search--show-action" style="background:#f2f2f2;">
      <i class="van-icon van-icon-search" style="color:undefined;">
        <!---->
      </i>
      <div class="van-cell van-field van-field--has-icon">
        <!---->
        <!---->
        <div class="van-cell__value van-cell__value--alone">
          <input type="search" value="" class="van-field__control">
          <!---->
          <div class="van-field__icon" style="display:none;">
            <i class="van-icon van-icon-clear" style="color:undefined;">
              <!---->
            </i>
          </div>
          <!---->
        </div>
        <!---->
      </div>
      <div class="van-search__action">
        <div>搜索</div>
      </div>
    </div>
  </section>
</section>
`;
